File: src/services/system.ts
import { invoke } from '@tauri-apps/api/core';

export interface SystemInfo {
  cpu_cores: number;
  os: string;
  arch: string;
}

let _cached: SystemInfo | null = null;

export async function getSystemInfo(): Promise<SystemInfo> {
  if (_cached) return _cached;
  _cached = await invoke<SystemInfo>('get_system_info');
  return _cached;
}

File: src-tauri/src/commands/system.rs
use serde::Serialize;

#[derive(Serialize)]
pub struct SystemInfo {
    /// Logical CPU cores (counts SMT threads). Useful for UX hints —
    /// hashing itself is single-threaded.
    pub cpu_cores: usize,
    pub os: String,
    pub arch: String,
}

#[tauri::command]
pub fn get_system_info() -> SystemInfo {
    let cpu_cores = std::thread::available_parallelism()
        .map(|n| n.get())
        .unwrap_or(1);
    SystemInfo {
        cpu_cores,
        os: std::env::consts::OS.to_string(),
        arch: std::env::consts::ARCH.to_string(),
    }
}
